### PR TITLE
EASY-1681: Return single FileInfo object for GET on specific file

### PR DESCRIPTION
**Pull request by vesaakerman**
Fixes EASY-1681

#### When applied it will
* return information about a file when the path in GET /deposit/{id}/file/{path} points to a file

#### Related pull requests on github
repo                       | PR                | note
-------------------------- | ----------------- | ----
easy-                      | [PR#](PRlink)     | ..

### CODE REVIEW

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 1]
```diff
@@ -61,6 +84,25 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
...
+   *
+   * @param path a relative path into a data files.
+   * @return a [[FileInfo]] object
+   */
+  def fileInfo(path: Path = Paths.get("")): Try[FileInfo] = {
+    val manifestMap = bag.payloadManifests
+    val absolutePath = dataDir / path.toString
+    val fileExists = manifestMap.get(SHA1).orElse(manifestMap.values.headOption).get.contains(absolutePath)
+    if (fileExists) {
+      val checksum = manifestMap.get(SHA1).orElse(manifestMap.values.headOption) get absolutePath
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: Questions:
* What will happen if there is no payload manifest at all?
* What are the invariants that we can count on?
* What do you report if an invariant is violated? 
* Would it make sense to check the file system to see if the file actually exists? What would you report if the file then exists on disk, but there is no manifest entry for it?

Comments:
* Lines 96 and 98 duplicate quite a lot.
* I see a lot of overlap with `fileInfoSeq`.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala [hunk 2]
```diff
@@ -191,22 +195,21 @@ class EasyDepositApiApp(configuration: Configuration) extends DebugEnhancedLoggi
...
    */
   def writeDataMetadataToDeposit(dm: DatasetMetadata, user: String, id: UUID): Try[Unit] = for {
-    deposit <- DepositDir.get(draftsDir, user, id)
+    deposit <- getDeposit(user, id)
     _ <- deposit.writeDatasetMetadataJson(dm)
   } yield ()
 
   /**
-   * Returns the list of [[FileInfo]] objects for the dataset files in this deposit.
+   * Returns dataset files of the deposit
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Why did you change the description?

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala [hunk 3]
```diff
@@ -119,8 +118,10 @@ class DepositServlet(app: EasyDepositApiApp)
     (for {
       uuid <- getUUID
       path <- getPath
-      depositFiles <- app.getDepositFiles(user.id, uuid, path)
-    } yield Ok(body = toJson(depositFiles))
+      dataFiles <- app.getDataFiles(user.id, uuid)
```
Comment (janvanmansum) [inline src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: This type of logic (deciding between directory? or single file?) should really be handled in the application object.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 4]
```diff
@@ -43,8 +66,8 @@ case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
    * @param path a relative path into data files directory of the bag.
    * @return a list of [[FileInfo]] objects
    */
-  def list(path: Path = Paths.get("")): Try[Seq[FileInfo]] = {
-    val parentPath = bag.baseDir / "data" / path.toString
+  def fileInfoSeq(path: Path = Paths.get("")): Try[Seq[FileInfo]] = {
```
[the remark this answers is not in the document]
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: By a request from @jo-pol , to show that the two methods (fileInfo and fileInfoSeq) are closely related.  I agree with Joke.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 5]
```diff
@@ -34,6 +35,28 @@ import scala.util.{ Failure, Success, Try }
...
+   * Returns 'true' if the path points to a directory.
+   *
+   * @param path a relative path.
+   * @return 'true' if directory, else 'false'
+   */
+  def isDirectory(path: Path): Boolean = {
+    dataDir / path.toString isDirectory
+  }
+
+  /**
```
[the remark this answers is not in the document]
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: It is used...

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 6]
```diff
@@ -34,6 +35,28 @@ import scala.util.{ Failure, Success, Try }
  */
 case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
 
+  val dataDir = bag.baseDir / "data"
```
[the remark this answers is not in the document]
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: Right, I've made it private now.

File: src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala (hunk 2 above)
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/EasyDepositApiApp.scala]: Because the method didn't return anymore [FileInfo] objects, but deposit data files.

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala (hunk 3 above)
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: I have now moved the distinction between directories and files into `EasyDepositApiApp`

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala (hunk 1 above)
Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: In my understanding the bag is already validated. It is possible still to check whether `manifestMap` is `null`, and if it is so, to throw an exception, and take this exception in account in `DepositServlet`.  But I am wondering if that is a bit overkill, if the deposit indeed has already been validated?  The same applies to the question whether we would check if the  file exists in the filesystem, even if it is absent in the manifest file.

Comment (vesaakerman) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: The lines 96 and 98 were indeed very similar, thus I modified the code a bit

**Comment by jo-pol** [comment]
I now see @janvanmansum asked for this change. https://github.com/DANS-KNAW/easy-deposit-api/pull/67#discussion_r216232451

The problem I tried to point out, according to the swagger spec the response body should be constructed with a `Seq[FileInfo].toJson` in case of a directory and a `FileInfo.toJson` in case of a file. So either the spec should change or we do need to call different methods from the app class. 

@lindareijnhoudt I don't know whether or how this is already implemented in the UI.

**Comment by jo-pol** [comment]
Pleas next time trim your post when replying via email, or adjust the recipient.
I now see you are returning and object, did not expect that.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala [hunk 7]
```diff
@@ -34,6 +35,28 @@ import scala.util.{ Failure, Success, Try }
  */
 case class DataFiles(bag: DansBag) extends DebugEnhancedLogging {
 
+  private val dataDir = bag.baseDir / "data"
```
Comment (jo-pol) [inline src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: This is the same as `bag.data`. I prefer the latter. so drop this one.

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala [hunk 8]
```diff
@@ -119,8 +118,9 @@ class DepositServlet(app: EasyDepositApiApp)
...
+      contents <- if (dataFiles.isDirectory(path)) dataFiles.list(path) else dataFiles.fileInfo(path)
```
[the remark this answers is not in the document]
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: @vesaakerman @jo-pol : I'd personally prefer just `list` and `get`. This short style was there from the beginning. Methods should normally have a name that is or starts with a verb.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: I found the comment with @jo-pol 's request. I disagree. I think we should try to keep the style uniform and also adhere to standard conventions in naming methods.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: I am referring to `fileContents` (which should by the way be called `getFileContents`). Where is it used, my IntelliJ can't locate the caller.

File: src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala (hunk 3 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/servlets/DepositServlet.scala]: @vesaakerman @jo-pol OK, that looks better. `getDataFiles` can now be made private by the way. It is however a pity that the return type of getFileInfo is now `Try[Object]`.  Here it *would* make sense to have sepate `listFileInfos` and `getFileInfo` methods.

File: src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala (hunk 7 above)
Comment (janvanmansum) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: @vesaakerman @jo-pol It is not a big deal to me, but there seems little value in this field. Any arguments in favor or contra (apart from preferences)?

Comment (jo-pol) [reply src/main/scala/nl.knaw.dans.easy.deposit/DataFiles.scala]: The field used to be valuable before the new bag-lib was used. Now that we also have `bag.data` it is confusing to use both. Only by this review I saw I should have cleaned it up when introducing the bag-lib.

File: src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala [hunk 9]
```diff
@@ -117,6 +115,31 @@ class IntegrationSpec extends TestSupportFixture with ServletFixture with Scalat
...
+
+    // get the file information
+    expectsUserFooBar
+    get(uri = s"/deposit/$uuid/file/path", headers = Seq(basicAuthentication)) {
+      status shouldBe OK_200
+      body shouldBe """[{"fileName":"text.txt","dirPath":"path/to","sha1sum":"c5b8de8cc3587aef4e118a481115391033621e06"}]"""
+    }
+  }
+
+  "scenario: POST /deposit; GET /deposit/:uuid/file/path/to/text.txt" should "return information of the file" in {
```
[the remark this answers is not in the document]
Comment (vesaakerman) [reply src/test/scala/nl.knaw.dans.easy.deposit/servlets/IntegrationSpec.scala]: I have reworded the comments;  I put the 'twice' in the right place, and also changed the text what the test should return.
